spark-bigquery-connector-common/src/test/java/com/google/cloud/spark/bigquery/integration/WriteIntegrationTestBase.java

```diff
@@ -1431,6 +1434,367 @@ public void testWriteStringToTimeField() throws Exception {
+            testDataset, testTable, orderId, orderDateTime));
+
+    Dataset<Row> df =
+        spark.createDataFrame(
+            Arrays.asList(
+                RowFactory.create(10, Timestamp.valueOf("2023-09-28 10:15:00")),
+                RowFactory.create(20, Timestamp.valueOf("2023-09-30 12:00:00"))),
+            structType(
+                StructField.apply(orderId, DataTypes.IntegerType, true, Metadata.empty()),
+                StructField.apply(orderDateTime, DataTypes.TimestampType, true, Metadata.empty())));
+
+    Dataset<Row> result = writeAndLoadDatasetOverwriteDynamicPartition(df);
+    assertThat(result.count()).isEqualTo(3);
+    List<Row> rows = result.collectAsList();
+
+    Row row1 = rows.get(0);
+    Row row2 = rows.get(1);
+    Row row3 = rows.get(2);
+
+    assertThat(row1.getLong(row1.fieldIndex(orderId))).isEqualTo(1);
+    assertThat(row1.getTimestamp(row1.fieldIndex(orderDateTime)))
+        .isEqualTo(Timestamp.valueOf("2023-09-28 1:00:00"));
+
+    assertThat(row2.getLong(row2.fieldIndex("order_id"))).isEqualTo(10);
+    assertThat(row2.getTimestamp(row2.fieldIndex(orderDateTime)))
+        .isEqualTo(Timestamp.valueOf("2023-09-28 10:15:00"));
+
+    assertThat(row3.getLong(row3.fieldIndex("order_id"))).isEqualTo(20);
+    assertThat(row3.getTimestamp(row3.fieldIndex(orderDateTime)))
+        .isEqualTo(Timestamp.valueOf("2023-09-30 12:00:00"));
+  }
+
+  @Test
+  public void testOverwriteDynamicPartition_partitionTimestampByDay() {
+    assumeThat(writeMethod, equalTo(WriteMethod.DIRECT));
+    String orderId = "order_id";
+    String orderDateTime = "order_date_time";
+    TimeZone.setDefault(TimeZone.getTimeZone("UTC"));
+    IntegrationTestUtils.runQuery(
+        String.format(
+            "CREATE TABLE `%s.%s` (%s INTEGER, %s TIMESTAMP) "
+                + "PARTITION BY DATE(order_date_time) "
+                + "AS SELECT * FROM UNNEST([(1, TIMESTAMP '2023-09-28 1:00:00 UTC'), "
+                + "(2, TIMESTAMP '2023-09-29 10:00:00 UTC'), (3, TIMESTAMP '2023-09-29 17:00:00 UTC')])",
+            testDataset, testTable, orderId, orderDateTime));
+
+    Dataset<Row> df =
+        spark.createDataFrame(
+            Arrays.asList(
+                RowFactory.create(10, Timestamp.valueOf("2023-09-29 2:00:00")),
+                RowFactory.create(20, Timestamp.valueOf("2023-09-30 12:00:00"))),
+            structType(
+                StructField.apply(orderId, DataTypes.IntegerType, true, Metadata.empty()),
+                StructField.apply(orderDateTime, DataTypes.TimestampType, true, Metadata.empty())));
+
+    Dataset<Row> result = writeAndLoadDatasetOverwriteDynamicPartition(df);
+    assertThat(result.count()).isEqualTo(3);
+    List<Row> rows = result.collectAsList();
+
+    Row row1 = rows.get(0);
+    Row row2 = rows.get(1);
+    Row row3 = rows.get(2);
+
+    assertThat(row1.getLong(row1.fieldIndex(orderId))).isEqualTo(1);
```

Comment:
You need to sort the result, order is not guaranteed

Reply:
Fixed